File: backend/app/schemas/jobs.py
from __future__ import annotations

from datetime import datetime
from typing import List, Literal

from pydantic import BaseModel, Field

JobTier = Literal["staging", "career"]
MatchSource = Literal["overlap", "llm", "none"]


class JobOfferRead(BaseModel):
    id: str
    title: str
    company: str
    url: str
    description: str
    required_skills: List[str]
    scraped_at: datetime
    # staging = scraped listing (Postgres); career = promoted JobOffer in the KG
    tier: JobTier = "career"
    location: str = ""
    salary_range: str | None = None
    job_type: str | None = None
    source: str | None = None
    seniority: str | None = None
    min_experience_years: int | None = None
    max_experience_years: int | None = None


class JobMatchResponse(BaseModel):
    job_offer: JobOfferRead
    score: float
    matching_skills: List[str] = Field(default_factory=list)
    missing_skills: List[str] = Field(default_factory=list)
    justification: str = ""
    quick_score: float = 0.0
    is_stale: bool = False
    matched_at: datetime | None = None
    source: MatchSource = "none"


class JobMatchBatchRequest(BaseModel):
    """Empty / omitted ``job_ids`` means all active career offers (capped)."""

    job_ids: list[str] | None = Field(
        default=None,
        description="Explicit job ids to LLM-match. Null/empty = all active career offers.",
    )


class JobPromoteResponse(BaseModel):
    job_offer: JobOfferRead
    promoted: bool = Field(
        description="True when a staging listing was newly promoted; False if already career"
    )

File: frontend/src/types/jobs.ts
export interface JobOffer {
  id: string;
  title: string;
  company: string;
  url: string;
  description: string;
  required_skills: string[];
  scraped_at: string;
  /** staging = scraped listing; career = saved into the personal KG */
  tier?: 'staging' | 'career';
  location?: string;
  salary_range?: string | null;
  job_type?: string | null;
  source?: string | null;
  seniority?: string | null;
  min_experience_years?: number | null;
  max_experience_years?: number | null;
}

export type ExperienceBracket = '0-2' | '2-5' | '5+';
export type SeniorityFilter = 'junior' | 'mid' | 'senior';
export type MatchSource = 'overlap' | 'llm' | 'none';

export interface ScrapedJobFilters {
  search?: string;
  source?: string;
  seniority?: SeniorityFilter | '';
  experience_bracket?: ExperienceBracket | '';
}

export interface JobMatch {
  job_offer: JobOffer;
  score: number;
  matching_skills?: string[];
  missing_skills?: string[];
  justification?: string;
  quick_score?: number;
  is_stale?: boolean;
  matched_at?: string | null;
  source?: MatchSource;
}

export interface JobPromoteResponse {
  job_offer: JobOffer;
  promoted: boolean;
}
